### ISSUE TITLE
Protobuf support doesn't work out-of-the-box…?

### ISSUE THREAD

**Issue by hedefalk** [zed-industries/zed#21057]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

It sounds to me that protobuf "should just work". Docs link to "protols" https://github.com/coder3101/protols while the code references:

protobuf-language-server:

https://github.com/zed-industries/zed/blob/114c4621433eb948b87fa5f4f41df8af6601f66e/extensions/proto/src/proto.rs

With this plugin installed I just get

```
Language server error: protobuf-language-server

protobuf-language-server not found in PATH
-- stderr--
```

I now installed a "protols" unto path with nix:

```
which protols
/run/current-system/sw/bin/protols
```

and stuck that into settings as:

```
"lsp": {
    ...
    "protobuf-language-server": {
      "binary": {
        "path": "protols"
      }
    }
}
```

Am I misunderstanding what's needed here? Otherwise, maybe that should go into the docs page over at https://zed.dev/docs/languages/proto

### Environment

Zed: v0.162.5 (Zed)
OS: macOS 15.1.0
Memory: 64 GiB
Architecture: aarch64

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```

```
</details>

**Comment by SturdyFool10** [comment]
its an extension called proto IIRC

**Comment by notpeter** [comment]
Yeah, docs need some love.  I tried to [add some formatting docs](https://github.com/zed-industries/zed/pull/20677) as the docs [very sparse](https://github.com/zed-industries/zed/blob/eb3c4b0e461886e87e6759471d69ea9a524a9d43/docs/src/languages/proto.md) but I incorrectly added the link to ProtoLS (oops).

Extension was originally written with protobuf-language-server in mind, protols is not super well tested.  I've commented out the part of the docs I'm less sure about until I have a chance to test further next week.

Thanks for reporting.

**Comment by scbizu** [comment]
Is there any option to use other lsp binary other than protols  (e.g.: bufls) ? Here is some context in https://github.com/zed-industries/extensions/issues/146

**Comment by notpeter** [comment]
I haven't tested, but others have and it should work. Specify bufls and the lsp binary or create a symlink in your path to bufls named "protobuf-language-server". Then specify "language_server" as the formatter.

https://zed.dev/docs/configuring-languages#configuring-formatters

**Comment by scbizu** [comment]
Yep ,  I correct my config with:

```json
"lsp": {
    "protobuf-language-server": {
      "binary": {
        "path": "buf",
        "arguments": ["beta", "lsp"]
      }
    }
}
```

And then the  `protobuf-language-server not found in PATH` do not show again

**Comment by orange-juzipi** [comment]
In proto, can we add a command to automatically sort the variable IDs in the message?

**Comment by notpeter** [comment]
> In proto, can we add a command to automatically sort the variable IDs in the message?

@orange-juzipi This is largely dependent on the language server exposing these a custom actions to Zed. You can see these via `editor: toggle code actions` (`cmd-.`).

For example a number of language servers expose a `organizeImports` action (gopls, ruff, etc) and you can even configure Zed to trigger those as part of formatting configuration. But there are many additional language-server specific actions which are available depending on context. For example [rust analyzer](https://rust-analyzer.github.io/manual.html#assists-code-actions) has like a hundred.  

[protobuf-language-server](https://github.com/lasorda/protobuf-language-server) does not implement any of these and appears not to be have much in the way of active development.

[bufbuild/buf](https://github.com/bufbuild/buf) is newer, but also doesn't have anything like this.

**Comment by MarkusPersad** [comment]
`"lsp": {
        "protobuf-language-server": {
            "binary": {
                "path": "/home/{UserName}/go/bin/protobuf-language-server"
            }
        }
    }`
When the binary path is configured to the absolute path of the protobuf-language-server, it works.

**Comment by andrewbanchich** [comment]
i've been running into this and it hasn't been clear how i should proceed.

all other extensions i've tried set up the language server themselves. why does this fail? why do i need to install it manually or mess around with configs / paths / other servers at all?

since `buf` seems to be the better maintained one, i tried the above config and it just crashes whenever i save. i'm assuming `buf` isn't fully compatible with `protobuf-language-server`?

**Comment by scbizu** [comment]
> since buf seems to be the better maintained one, i tried the above config and it just crashes whenever i save. i'm assuming buf isn't fully compatible with protobuf-language-server?

@andrewbanchich compatible , but not fully . It still in beta for years or two , if you'd like to try `buf` , you can use `buf beta` which I mentioned above . See [buf docs](https://buf.build/docs/reference/cli/buf/beta/lsp/) for details .

**Comment by andrewbanchich** [comment]
@scbizu i did try the config you provided, but it crashes when i save.

```
cannot format file "/path/to/schema.proto", 25 error(s) found
```

when i run `buf format schema.proto` in the CLI it does output the formatted text to stdout.

**Comment by scbizu** [comment]
@andrewbanchich yep , I tried with formatting , the result is the same as you . I have raised the issue to [buf](https://github.com/bufbuild/buf/issues/3701). You should try to fix the `buf lint` errors before formatting.

**Comment by sygi** [comment]
Would it be possible to add a bit more concrete docs about how to set this up? I understand one needs to install something and point the setting there, is this something protols? When doing remote development, I should install it on the remote server, right?

**Comment by Master-Y0da** [comment]
can't format any proto file...same error as @andrewbanchich

**Comment by Jason5Lee** [comment]
> Yep , I correct my config with:
> 
> "lsp": {
>     "protobuf-language-server": {
>       "binary": {
>         "path": "buf",
>         "arguments": ["beta", "lsp"]
>       }
>     }
> }
> And then the `protobuf-language-server not found in PATH` do not show again

I try this, then I get the following warning on every proto file.

`Files with package "<package name>" must be within a directory "<package name>" relative to root but were in directory ".".`

It seems it cannot detect the directory.

**Comment by warrenjokinen** [comment]
**Interesting**, but not exactly a duplicate of my  #40058 

I don't see here any mention of Proto v0.2.2 Extension (from Zed Industries)
I don't know if the extension is obsolete, or ?

Despite that this issue here has been Open for almost a year, I still don't know how to proceed.

**Comment by scbizu** [comment]
@warrenjokinen I am using Proto v0.2.2 Extension , Proto Extension allow users to set various LSP servers via `protobuf-language-server` setting section, such as [protols](https://github.com/coder3101/protols) , [buf lsp](https://github.com/bufbuild/buf/tree/main/private/buf/buflsp) , which Zed extension provided is just a LSP client to interact with these servers via language server protocol. 

So , in my opinion , the bug maybe hard to resolved since it depends on 3rd tools , the actual behavior maybe different though. 

You can check [the extension source code](https://github.com/zed-industries/zed/blob/main/extensions/proto/src/proto.rs#L18-L40) for more details .

(BTW , for highlighting, I think the tree-sitter plugin is stable now)

**Comment by icdevin** [comment]
So, is this not expected to just work out-of-the-box? I feel like it should? I don't think I needed to install any other language servers for other language extensions to work (but it's possible they were already available on my system?)

I did get it working with `bufls`, just feels like that shouldn't be necessary.
